Plutarch/List.hs

```diff
@@ -142,23 +147,41 @@ pelem =
 -- | / O(n) /. Count the number of elements in the list
 plength :: PIsListLike list a => Term s (list a :--> PInteger)
 plength = phoistAcyclic $
-  plet
-    ( pfix #$ plam $ \self ls n ->
-        pelimList
-          (\_ xs -> self # xs # n + 1)
-          n
-          ls
-    )
-    $ \go -> plam $ \xs -> go # xs # 0
+  plam $ \xs ->
+    let go :: PIsListLike list a => Term s (list a :--> PInteger :--> PInteger)
+        go = (pfix #$ plam $ \self ls n -> pelimList (\_ xs -> self # xs # n + 1) n ls)
+     in go # xs # 0
 
 --------------------------------------------------------------------------------
 
--- | / O(n) /. Fold on a list right-associatively
-pfoldr :: PIsListLike list a => Term s ((a :--> b :--> b) :--> b :--> list a :--> b)
+-- | / O(n) /. Fold on a list left-associatively
+pfoldl :: PIsListLike list a => Term s ((b :--> a :--> b) :--> b :--> list a :--> b)
+pfoldl = phoistAcyclic $
+  plam $ \f ->
+    pfix #$ plam $ \self z l ->
+      pelimList
+        (\x xs -> self # (f # z # x) # xs)
+        z
+        l
+
+-- | The same as 'pfoldl', but with Haskell-level reduction function.
+pfoldl' :: PIsListLike list a => (forall s. Term s b -> Term s a -> Term s b) -> Term s (b :--> list a :--> b)
+pfoldl' f = phoistAcyclic $
+  pfix #$ plam $ \self z l ->
+    pelimList
+      (\x xs -> self # f z x # xs)
+      z
+      l
+
+{- | / O(n) /. Fold on a list right-associatively.
+
+May short circuit if given reducer function is lazy in its second argument.
+-}
+pfoldr :: PIsListLike list a => Term s ((a :--> PDelayed b :--> b) :--> b :--> list a :--> b)
```

Comment:
Assume we have an existing use of `pfoldr` where we e.g. multiply a list of quaternions, so it's dependent on the order. We would immediately force the accumulator no matter what, and hence it would be no more efficient than what we have now. We probably want another function that has a delayed accumulator rather than changing `pfoldr`.

Reply:
Good idea! I was considering that as well. Suggestion on naming?

Reply:
`pfoldrLazy`?

Reply:
Done!